### ISSUE TITLE
calculate_targets SimpleNNQPlayer.py

### ISSUE THREAD

**Issue by PaulBunkerWhich** [fcarsten/tic-tac-toe#2]
Hi,

Me again! I have a query about the calculate_targets method in the SimpleNNQPlayer class.

It seems to me that we are only really teaching the model about final moves.

In TQPlayer when we loop through the moves at the end of the game we reverse the game moves and apply the formula to set the new Q value for the final move first. Then we set the next_max for us to use on the next iteration of the loop, the previous move. Here it makes sense that we use the reward discount to reduce the penalty/reward in steps and so on until the first move.

However

When we call calculate_targets for the NNQ player 

```python
target[self.action_log[i]] = self.reward_discount * self.next_max_log[i]
````

self.next_max_log[i] is just some arbitrary value given to us by a naive neural network and only when we get to the final iteration of the loop where self.next_max_log[i] is the actual reward/punishment does this calculation seem to make sense. The next_max values aren't related to the result of the game.

It feels to me like we should be doing something here more akin to what is going on in the TabQ player reward mechanism?

Thanks for reading
-Paul

**Comment by fcarsten** [comment]
Hi Paul,

`self.reward_discount` is a misleading variable name, and I need to change it. It is really just the discount factor from the tabular Q function case. The "reward/punishment" terminology is also probably not the right one to use here, and more relevant later on when we look at Policy Gradient methods.

With that in mind, the NN more or less does what we did in the tabular Q case: For each state and move we try to train it to predict the maximum Q value of the next state. Unfortunately, like in the tabular case, we do not know what the real maximum Q value of the next state is, so we have to learn that at the same time. 

`self.next_max_log` is the NN's current best guess at what the max Q value of the next state is, similar to the corresponding max Q value in the tabular approach. `self.reward_discount` is the discount factor, also similar to the tabular case. By setting the product of `self.reward_discount` and `self.next_max_log` as the target for the NN and performing the gradient descent, we make the NN change its current value slightly towards the target value. This is the equivalent to the `Q(S,A)=(1−α)∗Q(S,A)+α∗γ∗maxaQ(S′,a)` update from the tabular approach.

Apart from the very last move, the `self.next_max_log[i] ` values are indeed, at least in the beginning, arbitrary. Due to the last move values being correct though, the NN will however, be at least a little bit more correct after the first iteration. In particular it will be better at predicting the Q value of the last move. This effect will trickle down to earlier moves in the game eventually by playing over and over again, and eventually converge on the true Q function for all states and moves - or at least so we hope.

**Comment by PaulBunkerWhich** [comment]
AHA! 

So If my understanding is correct... 

We win the first game by placing an x in the top right corner and we train the NN to update the Q-value  for that move given the previous board state. The other iterations of the loop at this stage are arbitrary and not particularly help-full at this stage. 

The second game by chance both players play the exact same moves, the penultimate iteration of the loop will have the q values for the next state where we saw the winning move and have a high Q value in that array (although not in the position we are placing our token) we apply the ```reward_discount``` / ```reward_decay``` to this high Q-value and pop it in the place of our move.

I see now how the trickle down effect works with more games.

Thanks for the explanation this is sitting much better with me now. 

-Paul

**Comment by fcarsten** [comment]
Yep, that's it! 

I'll close the issue once I have renamed the discount variable to something more appropriate like `discount_factor` (or if you have already done so in your version, feel free to submit a pull request and I'll merge it in).

**Comment by PaulBunkerWhich** [comment]
I would have loved to submit a pull request but my code is unfortunately in javascript. :/ 

[https://github.com/PaulBunker/Naughts-and-Crosses/tree/master/src](url)

It's a bit of a struggle trying to implement some of the model architecture as the API is different but I'll get there eventually. I have similar results now with SimpleNNQPlayer and working on the next step!

Thanks again 
-Paul

**Comment by fcarsten** [comment]
No worries and great idea to try doing it in another programming language! Let me know if you manage to get to a full implementation.
